arangod/Replication2/ReplicatedLog/InMemoryLog.cpp

```diff
@@ -0,0 +1,313 @@
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#include "InMemoryLog.h"
+
+#include "Replication2/LoggerContext.h"
+#include "Replication2/ReplicatedLog/LogCore.h"
+#include "Replication2/ReplicatedLog/PersistedLog.h"
+#include "Replication2/ReplicatedLog/ReplicatedLogIterator.h"
+
+#include <Basics/Exceptions.h>
+#include <Basics/StringUtils.h>
+#include <Basics/application-exit.h>
+#include <Basics/debugging.h>
+#include <Containers/ImmerMemoryPolicy.h>
+
+#if (_MSC_VER >= 1)
+// suppress warnings:
+#pragma warning(push)
+// conversion from 'size_t' to 'immer::detail::rbts::count_t', possible loss of data
+#pragma warning(disable : 4267)
+// result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)
+#pragma warning(disable : 4334)
+#endif
+#include <immer/flex_vector.hpp>
+#include <immer/flex_vector_transient.hpp>
+#if (_MSC_VER >= 1)
+#pragma warning(pop)
+#endif
+
+using namespace arangodb;
+using namespace arangodb::replication2;
+
+auto replicated_log::InMemoryLog::getLastIndex() const noexcept -> LogIndex {
+  auto const result = LogIndex{_log.size()};
+  // log empty => result == 0
+  TRI_ASSERT(!_log.empty() || result == LogIndex(0));
+  // !log empty => result index == last entry
+  TRI_ASSERT(_log.empty() || result == _log.back().entry().logIndex());
+  return result;
+}
+
+auto replicated_log::InMemoryLog::getLastTermIndexPair() const noexcept -> TermIndexPair{
+  if (_log.empty()) {
+    return {};
+  }
+  return _log.back().entry().logTermIndexPair();
+}
+
+auto replicated_log::InMemoryLog::getLastTerm() const noexcept -> LogTerm {
+  if (_log.empty()) {
+    return LogTerm{0};
+  }
+  return _log.back().entry().logTerm();
+}
+
+auto replicated_log::InMemoryLog::getNextIndex() const noexcept -> LogIndex {
+  return getLastIndex() + 1;
+}
+
+auto replicated_log::InMemoryLog::getEntryByIndex(LogIndex const idx) const noexcept
+    -> std::optional<InMemoryLogEntry> {
+  if (_log.size() < idx.value || idx.value == 0) {
+    return std::nullopt;
+  }
+
+  auto const& e = _log.at(idx.value - 1);
+  TRI_ASSERT(e.entry().logIndex() == idx);
+  return e;
+}
+
+auto replicated_log::InMemoryLog::splice(LogIndex from, LogIndex to) const -> log_type {
```

Comment:
I think the name is misleading, this seems to get a "slice", not do any kind of "splicing".